// @soundworks/core/common/SharedState.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _parameters = _interopRequireDefault(require("@ircam/parameters"));

var _lodash = _interopRequireDefault(require("lodash.clonedeep"));

var _sharedStateUtils = require("./shared-state-utils.js");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// -----------------------------------------------------------
// define a common namespace...
// -----------------------------------------------------------

/**
 * Representation of a shared state. The class and its instances are the same
 * client-side and server-side.
 *
 * @memberof common
 *
 * @see {client.SharedStateManagerClient}
 * @see {server.SharedStateManagerServer}
 */
class SharedState {
  constructor(id, remoteId, schemaName, schema, client, isCreator, manager, initValues = {}) {
    this.id = id;
    this.remoteId = remoteId;
    this.schemaName = schemaName;
    this._schema = (0, _lodash.default)(schema);
    this._isCreator = isCreator; // may be the server or any client

    this._client = client;
    this._manager = manager;

    try {
      this._parameters = (0, _parameters.default)(schema, initValues);
    } catch (err) {
      console.error(err.stack);
      throw new Error(`Error creating or attaching state "${schemaName}" w/ values:\n
${JSON.stringify(initValues, null, 2)}`);
    }

    this._subscriptions = new Set();
    this._onDetachCallbacks = new Set();
    this._onDeleteCallbacks = new Set(); // add listener for state updates

    client.transport.addListener(`${_sharedStateUtils.UPDATE_RESPONSE}-${id}-${this.remoteId}`, (reqId, updates) => {
      const updated = this._commit(updates);

      (0, _sharedStateUtils.resolveRequest)(reqId, updated);
    }); // retrieve values but do not propagate to subscriptions

    client.transport.addListener(`${_sharedStateUtils.UPDATE_ABORT}-${id}-${this.remoteId}`, (reqId, updates) => {
      const updated = this._commit(updates, false);

      (0, _sharedStateUtils.resolveRequest)(reqId, updated);
    });
    client.transport.addListener(`${_sharedStateUtils.UPDATE_NOTIFICATION}-${id}-${this.remoteId}`, updates => {
      this._commit(updates);
    }); // ---------------------------------------------
    // DELETE initiated by creator, or schema deleted
    // ---------------------------------------------

    client.transport.addListener(`${_sharedStateUtils.DELETE_NOTIFICATION}-${id}-${this.remoteId}`, () => {
      this._manager._statesById.delete(this.id);

      this._clearTransport();

      this._onDetachCallbacks.forEach(callback => callback());

      this._onDeleteCallbacks.forEach(callback => callback());
    });

    if (this._isCreator) {
      // ---------------------------------------------
      // DELETE (can only delete if creator)
      // ---------------------------------------------
      client.transport.addListener(`${_sharedStateUtils.DELETE_RESPONSE}-${id}-${this.remoteId}`, reqId => {
        this._manager._statesById.delete(this.id);

        this._clearTransport();

        this._onDetachCallbacks.forEach(callback => callback());

        this._onDeleteCallbacks.forEach(callback => callback());

        (0, _sharedStateUtils.resolveRequest)(reqId, this);
      });
      client.transport.addListener(`${_sharedStateUtils.DELETE_ERROR}-${id}`, (reqId, msg) => {
        (0, _sharedStateUtils.rejectRequest)(reqId, msg);
      });
    } else {
      // ---------------------------------------------
      // DETACH
      // ---------------------------------------------
      client.transport.addListener(`${_sharedStateUtils.DETACH_RESPONSE}-${id}-${this.remoteId}`, reqId => {
        this._manager._statesById.delete(this.id);

        this._clearTransport();

        this._onDetachCallbacks.forEach(func => func());

        (0, _sharedStateUtils.resolveRequest)(reqId, this);
      }); // the state does not exists anymore in the server (should not happen)

      client.transport.addListener(`${_sharedStateUtils.DETACH_ERROR}-${id}`, (reqId, msg) => {
        (0, _sharedStateUtils.rejectRequest)(reqId, msg);
      });
    }
  }

  _clearTransport() {
    // remove listeners
    this._client.transport.removeAllListeners(`${_sharedStateUtils.UPDATE_RESPONSE}-${this.id}-${this.remoteId}`);

    this._client.transport.removeAllListeners(`${_sharedStateUtils.UPDATE_NOTIFICATION}-${this.id}-${this.remoteId}`);

    this._client.transport.removeAllListeners(`${_sharedStateUtils.DELETE_RESPONSE}-${this.id}-${this.remoteId}`);

    this._client.transport.removeAllListeners(`${_sharedStateUtils.DELETE_NOTIFICATION}-${this.id}-${this.remoteId}`);

    this._client.transport.removeAllListeners(`${_sharedStateUtils.DETACH_RESPONSE}-${this.id}-${this.remoteId}`);
  }
  /** @private */


  _commit(obj, propagate = true) {
    const updated = {};

    for (let name in obj) {
      updated[name] = this._parameters.set(name, obj[name]);
    } // if the `UPDATE_REQUEST` as been aborted by the server, do not propagate


    if (propagate) {
      this._subscriptions.forEach(listener => listener(updated));
    }

    return updated;
  }
  /**
   * Get the schema that describes the state.
   *
   * @return {Object}
   */


  getSchema() {
    return this._schema;
  }
  /**
   * Updates values of the state.
   *
   * @async
   * @param {Object} updates - key / value pairs of updates to apply to the state
   * @return {Object}
   */


  async set(updates) {
    return new Promise((resolve, reject) => {
      const reqId = (0, _sharedStateUtils.storeRequestPromise)(resolve, reject);

      this._client.transport.emit(`${_sharedStateUtils.UPDATE_REQUEST}-${this.id}-${this.remoteId}`, reqId, updates);
    });
  }
  /**
   * Get a value of the state by its name
   *
   * @param {String} name - Name of the param
   * @return {Mixed}
   */


  get(name) {
    return this._parameters.get(name);
  }
  /**
   * Get a all the key / value pairs of the state.
   *
   * @return {Object}
   */


  getValues() {
    return this._parameters.getValues();
  }
  /**
   * @callback common.SharedState~subscribeCallback
   * @param {Object} updates - key / value pairs of the updates that have been
   *   applied to the state
   */

  /**
   * Subscribe to state updates
   *
   * @param {common.SharedState~subscribeCallback} callback - callback to execute
   *   when an update is commited on the state.
   * @example
   * state.subscribe(async (updates) =>  {
   *   for (let [key, value] of Object.entries(updates)) {
   *      // dispatch
   *   }
   * }
   */


  subscribe(listener) {
    this._subscriptions.add(listener);

    return () => {
      this._subscriptions.delete(listener);
    };
  }
  /**
   * Detach from the state. If the client is the creator of the state, the state
   * is deleted and all attached nodes get notified
   *
   * @async
   * @see {common.SharedState#onDetach}
   * @see {common.SharedState#onDelete}
   * @see {client.SharedStateManagerClient#create}
   * @see {server.SharedStateManagerClient#create}
   * @see {client.SharedStateManagerClient#attach}
   * @see {server.SharedStateManagerClient#attach}
   */


  async detach() {
    this._subscriptions.clear();

    if (this._isCreator) {
      return new Promise((resolve, reject) => {
        const reqId = (0, _sharedStateUtils.storeRequestPromise)(resolve, reject);

        this._client.transport.emit(`${_sharedStateUtils.DELETE_REQUEST}-${this.id}-${this.remoteId}`, reqId);
      });
    } else {
      return new Promise((resolve, reject) => {
        const reqId = (0, _sharedStateUtils.storeRequestPromise)(resolve, reject);

        this._client.transport.emit(`${_sharedStateUtils.DETACH_REQUEST}-${this.id}-${this.remoteId}`, reqId);
      });
    }
  }
  /**
   * Register a function to execute when detaching from the state
   *
   * @param {Function} callback - callback to execute when detaching from the state.
   *   wether the client as called `detach`, or the state has been deleted by its
   *   creator.
   */


  onDetach(callback) {
    this._onDetachCallbacks.add(callback);

    return () => this._onDetachCallbacks.delete(callback);
  }
  /**
   * Register a function to execute when the state is deleted. Only called if the
   * node was the creator of the state. Is called after `onDetach`
   *
   * @param {Function} callback - callback to execute when the state is deleted.
   */


  onDelete(callback) {
    this._onDeleteCallbacks.add(callback);

    return () => this._onDeleteCallbacks.delete(callback);
  }

}

var _default = SharedState;
exports.default = _default;
